spec/rubocop/cop/rspec/leaked_const_spec.rb

```diff
@@ -0,0 +1,100 @@
+RSpec.describe RuboCop::Cop::RSpec::LeakedConst do
+  subject(:cop) { described_class.new }
+
+  it 'finds a new named class defined in-line' do
+    expect_violation(<<-RUBY)
+      class Foo
+      ^^^^^^^^^ Ensure that new constants are cleaned up between specs
+      end
+    RUBY
+  end
+
+  it 'finds a new named class defined inside of a let block' do
+    expect_violation(<<-RUBY)
+      let(:fake_class) do
+        class Foo
+        ^^^^^^^^^ Ensure that new constants are cleaned up between specs
+        end
+      end
+    RUBY
+  end
+
+  it 'finds a new named class assigned to a variable' do
+    expect_violation(<<-RUBY)
+      fake_class = class Foo
+                   ^^^^^^^^^ Ensure that new constants are cleaned up between specs
+      end
+    RUBY
+  end
+
+  it 'finds a new unnamed class inside of a let block' do
+    expect_no_violations(<<-RUBY)
+      let(:fake_class) { Class.new }
+    RUBY
+  end
+
+  it 'finds a new unnamed class assigned to a variable' do
+    expect_no_violations(<<-RUBY)
+      fake_class = Class.new
+    RUBY
+  end
+
+  it 'finds a new class already marked with stub_const' do
+    expect_no_violations(<<-RUBY)
+      before do
+        stub_const('Foo', Class.new)
+        class Foo
+        end
+      end
+    RUBY
+  end
+
+  it 'finds a new class in a module that is already marked with stub_const' do
+    expect_no_violations(<<-RUBY)
+      before do
+        stub_const('Namespace', Module.new)
+        module Namespace
+          class Foo
+          end
+        end
+      end
+    RUBY
+  end
+
+  it 'finds two new classes in a module already marked with stub_const' do
+    expect_no_violations(<<-RUBY)
+      before do
+        stub_const('Namespace::SecondNamespace', Module.new)
+        module Namespace
+          module SecondNamespace
+            class Foo
+            end
+
+            class Bar
+            end
+          end
+        end
+      end
+    RUBY
+  end
+
+  it 'finds a new class marked for stub_const after opening the class' do
+    expect_violation(<<-RUBY)
+      before do
+        class Foo
+        ^^^^^^^^^ Ensure that new constants are cleaned up between specs
+        end
+        stub_const('Foo', Class.new)
+      end
+    RUBY
+  end
+
+  it 'finds a basic test set-up using modules to open the describe block' do
+    expect_no_violations(<<-RUBY)
+      module TestedNamespace
+        describe Subject do
+        end
+      end
+    RUBY
+  end
+end
```

Comment:
Another failing test case you may find helpful

```ruby
  it 'handles separate stubbing and class creation' do
    expect_no_violations(<<-RUBY)
      RSpec.describe Subject do
        before { stub_const('Foo', Class.new) }
        
        it 'does stuff' do
          class Foo
            def thing
            end
          end
          
          expect(Foo.new.thing).to be(nil)
        end
      end
    RUBY
  end
```

I would also suggest trying cases where you are in different `describe`/`context` nestings.